infrastructure/src/main/java/org/corfudb/infrastructure/SequencerServerCache.java

```diff
@@ -126,56 +81,76 @@ public void delete(@Nonnull ConflictTxStream key, Long globalAddress, @Nonnull R
      * @return global address
      */
     public Long getIfPresent(ConflictTxStream conflictKey) {
-        return conflictCache.getIfPresent(conflictKey);
+        return cacheConflictKeys.get(conflictKey);
+    }
+
+    /**
+     * The first address in the priority queue.
+     */
+    public long firstAddress() {
+        if (cacheEntries.isEmpty())
+            return Address.NOT_FOUND;
+
+        return cacheEntries.peek().txVersion;
+    }
+
+    /**
+     * Invalidate the records with the minAddress. It could be one or multiple records
+     */
+    private void invalidateFirst() {
+        ConflictTxStream firstEntry = cacheEntries.peek();
+        if (firstEntry == null) {
+            return;
+        }
+
+        int i = 0;
+        while (!cacheEntries.isEmpty() && cacheEntries.peek().txVersion == firstEntry.txVersion) {
+            ConflictTxStream entry = cacheEntries.poll();
+            cacheConflictKeys.remove(entry);
+            i++;
+        }
+
+        log.debug("Evict {} entries", i);
+        maxConflictWildcard = Math.max(maxConflictWildcard, firstEntry.txVersion);
     }
 
     /**
-     * Invalidate all records up to a trim mark.
+     * Invalidate all records up to a trim mark (not included).
      *
      * @param trimMark trim mark
      */
     public void invalidateUpTo(long trimMark) {
         log.debug("Invalidate sequencer cache. Trim mark: {}", trimMark);
+        int entries = 0;
 
-        AtomicLong entries = new AtomicLong();
-
-        conflictCache.asMap().forEach((key, txVersion) -> {
-            if (txVersion == null || txVersion >= trimMark) {
-                return;
-            }
-
-            conflictCache.invalidate(key);
-            entries.incrementAndGet();
-        });
-
-        log.info("Invalidated entries: {}", entries.get());
+        while(Address.isAddress(firstAddress()) && firstAddress() < trimMark) {
+            invalidateFirst();
+            entries++;
```

Comment:
Is the counter entries accurate? because invalidateFirst() might remove multiple entries.

Reply:
Good point @zhangn49 

invalidateFirst can return number of entries evicted and then that can be used in the counter.

Reply:
Well if the entries refer to the sequence number, then this counter is correct. If they strictly refer to cache entries, then it isnt correct because it under counts.